commit/plugin.go

```diff
@@ -85,8 +102,147 @@ func NewPlugin(
+		merkleProcessor:   merkleProcessor,
+		tokenProcessor:    tokenprice.NewProcessor(),
+		feeProcessor:      chainfee.NewProcessor(),
+	}
+}
+
+func (p *Plugin) Query(_ context.Context, outCtx ocr3types.OutcomeContext) (types.Query, error) {
+	return types.Query{}, nil
+}
+
+func (p *Plugin) ObservationQuorum(_ ocr3types.OutcomeContext, _ types.Query) (ocr3types.Quorum, error) {
+	// Across all chains we require at least 2F+1 observations.
+	return ocr3types.QuorumTwoFPlusOne, nil
+}
+
+func (p *Plugin) Observation(
+	ctx context.Context, outCtx ocr3types.OutcomeContext, _ types.Query,
+) (types.Observation, error) {
+	prevOutcome := p.decodeOutcome(outCtx.PreviousOutcome)
+	fChain := p.ObserveFChain()
+	//TODO: Send fchain instead of computing it inside
+	merkleObs, err := p.merkleProcessor.Observation(ctx, prevOutcome.MerkleRootOutcome, merkleroot.Query{})
+	if err != nil {
+		p.lggr.Errorw("failed to get merkle observation", "err", err)
+	}
+	tokenPricesObs, err := p.tokenProcessor.Observation(ctx, prevOutcome.TokenPriceOutcome, tokenprice.Query{})
+	if err != nil {
+		//log error
+		p.lggr.Errorw("failed to get token prices", "err", err)
+	}
+	gasObs, err := p.feeProcessor.Observation(ctx, prevOutcome.ChainFeeOutcome, chainfee.Query{})
+	if err != nil {
+		p.lggr.Errorw("failed to get gas prices", "err", err)
 	}
+
+	obs := Observation{
+		MerkleObs: merkleObs,
+		TokenObs:  tokenPricesObs,
+		GasObs:    gasObs,
+		FChain:    fChain,
+	}
+	return obs.Encode()
+}
+
+func (p *Plugin) ObserveFChain() map[cciptypes.ChainSelector]int {
+	fChain, err := p.homeChain.GetFChain()
+	if err != nil {
+		// TODO: metrics
+		p.lggr.Warnw("call to GetFChain failed", "err", err)
+		return map[cciptypes.ChainSelector]int{}
+	}
+	return fChain
+}
+
+// Outcome depending on the current state, either:
+// - chooses the seq num ranges for the next round
+// - builds a report
+// - checks for the transmission of a previous report
+func (p *Plugin) Outcome(
+	outCtx ocr3types.OutcomeContext, query types.Query, aos []types.AttributedObservation,
+) (ocr3types.Outcome, error) {
+	prevOutcome := p.decodeOutcome(outCtx.PreviousOutcome)
+
+	var merkleObservations []MerkleRootObservation
+	var tokensObservations []TokenPricesObservation
+	var feeObservations []ChainFeeObservation
+
+	for _, ao := range aos {
+		obs, err := DecodeCommitPluginObservation(ao.Observation)
+		if err != nil {
+			p.lggr.Errorw("failed to decode observation", "err", err)
+			continue
+		}
+		merkleObservations = append(merkleObservations,
+			MerkleRootObservation{
+				OracleID:    ao.Observer,
+				Observation: obs.MerkleObs,
+			},
+		)
+
+		tokensObservations = append(tokensObservations,
+			TokenPricesObservation{
+				OracleID:    ao.Observer,
+				Observation: obs.TokenObs,
+			},
+		)
+
+		feeObservations = append(feeObservations,
+			ChainFeeObservation{
+				OracleID:    ao.Observer,
+				Observation: obs.GasObs,
+			},
+		)
+	}
+
+	wg := &sync.WaitGroup{}
+	wg.Add(3)
```

Comment:
This code below feels like premature optimization. `Outcome` should be pure and "instant" so this concurrency probably adds more overhead than it gets rid of; benchmarking would be needed to see if thats the case.